### ISSUE TITLE
[加密货币恐慌指数策略] | Crypto Fear Greed Index

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4472]
**标签**: quantitative-strategy
### 摘要
基于加密货币市场情绪构建恐慌/贪婪指数，通过情绪极值反转策略捕捉市场拐点，实现高胜率的逆向交易。

### 算法描述

**核心原理：**

加密货币市场情绪指数（Fear & Greed Index）综合反映市场恐惧与贪婪程度。当市场极度恐惧时，价格往往被过度低估，是买入机会；当市场极度贪婪时，价格往往被过度高估，是卖出机会。情绪指数遵循均值回复规律，极端值后市场反转概率显著提高。

**数学模型：**

1. **情绪指数构建**：
$$FGI_t = \sum_{i=1}^{n} w_i \cdot S_i(t)$$

其中 $S_i(t)$ 为各情绪因子标准化得分，$w_i$ 为权重。

2. **情绪因子计算**：

- **波动率因子**（25%权重）：
$$S_{vol} = \frac{VIX_t - VIX_{min}}{VIX_{max} - VIX_{min}} \cdot 100$$

- **市场动量因子**（25%权重）：
$$S_{mom} = \frac{R_t - R_{min}}{R_{max} - R_{min}} \cdot 100$$

- **社交媒体因子**（15%权重）：
$$S_{social} = \frac{SI_t - SI_{min}}{SI_{max} - SI_{min}} \cdot 100$$

- **调查问卷因子**（15%权重）：
$$S_{survey} = \frac{BS_t - BS_{min}}{BS_{max} - BS_{min}} \cdot 100$$

- **Dominance因子**（10%权重）：
$$S_{dom} = \frac{DOM_t - DOM_{min}}{DOM_{max} - DOM_{min}} \cdot 100$$

- **搜索趋势因子**（10%权重）：
$$S_{search} = \frac{G_trend_t - G_{min}}{G_{max} - G_{min}} \cdot 100$$

3. **极值信号**：
$$FGI_t < 20 \Rightarrow \text{极度恐惧} \Rightarrow \text{买入信号}$$
$$FGI_t > 80 \Rightarrow \text{极度贪婪} \Rightarrow \text{卖出信号}$$

4. **Z-score偏离**：
$$z_t = \frac{FGI_t - \mu_{FGI}}{\sigma_{FGI}}$$

当 $z_t < -1.5$ 时市场情绪极端偏离均值，后续反转概率超过65%。

5. **RSI确认**：
$$RSI = 100 - \frac{100}{1 + RS}$$
$$RS = \frac{\text{平均上涨}}{\text{平均下跌}}$$

**交易策略：**

1. **极值反转策略**：FGI < 25 且 RSI < 35 → 买入；FGI > 75 且 RSI > 65 → 卖出
2. **趋势跟踪策略**：FGI 从低位突破 50 均线时追多；从高位跌破时追空
3. **背离策略**：价格创新低但 FGI 未创新低 → 看涨背离买入

### Python代码示例

```python
"""
加密货币恐慌指数策略
Crypto Fear Greed Index Strategy

核心逻辑：
1. 构建多因子情绪指数
2. 检测极端情绪极值
3. 配合RSI确认信号
4. 执行逆向交易

作者：Ginkgo量化研究组
版本：v1.0
"""

import numpy as np
import pandas as pd
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
from enum import Enum


class SentimentLevel(Enum):
    """情绪等级"""
    EXTREME_FEAR = "extreme_fear"
    FEAR = "fear"
    NEUTRAL = "neutral"
    GREED = "greed"
    EXTREME_GREED = "extreme_greed"


@dataclass
class FearGreedIndex:
    """恐慌贪婪指数数据"""
    timestamp: datetime
    value: float
    level: SentimentLevel
    volatility_score: float
    momentum_score: float
    social_score: float
    survey_score: float
    dominance_score: float
    search_score: float


class CryptoFearGreedIndexCalculator:
    """加密货币恐慌指数计算器"""
    
    def __init__(
        self,
        volatility_weight: float = 0.25,
        momentum_weight: float = 0.25,
        social_weight: float = 0.15,
        survey_weight: float = 0.15,
        dominance_weight: float = 0.10,
        search_weight: float = 0.10
    ):
        """
        初始化计算器
        
        Args:
            volatility_weight: 波动率权重
            momentum_weight: 动量权重
            social_weight: 社交媒体权重
            survey_weight: 调查问卷权重
            dominance_weight: Dominance权重
            search_weight: 搜索趋势权重
        """
        self.weights = {
            'volatility': volatility_weight,
            'momentum': momentum_weight,
            'social': social_weight,
            'survey': survey_weight,
            'dominance': dominance_weight,
            'search': search_weight
        }
        
        self.normalize_window = 30  # 标准化窗口
    
    def normalize_score(self, value: float, history: pd.Series) -> float:
        """
        将原始值标准化到0-100区间
        
        Args:
            value: 原始值
            history: 历史数据
        
        Returns:
            标准化得分
        """
        if len(history) < self.normalize_window:
            min_val, max_val = history.min(), history.max()
        else:
            min_val, max_val = history.iloc[-self.normalize_window:].min(), \
                             history.iloc[-self.normalize_window:].max()
        
        if max_val == min_val:
            return 50.0
        
        score = (value - min_val) / (max_val - min_val) * 100
        return np.clip(score, 0, 100)
    
    def calculate_fear_greed_index(
        self,
        volatility: float,
        momentum: float,
        social_mentions: int,
        bull_sentiment: float,
        btc_dominance: float,
        google_trends: float,
        history: Optional[Dict[str, pd.Series]] = None
    ) -> FearGreedIndex:
        """
        计算恐慌指数
        
        Args:
            volatility: 波动率指标
            momentum: 动量指标
            social_mentions: 社交媒体提及数
            bull_sentiment: 看涨情绪比例
            btc_dominance: BTC主导地位
            google_trends: Google趋势指数
            history: 历史数据用于标准化
        
        Returns:
            恐慌指数数据
        """
        timestamp = datetime.now()
        
        # 各因子得分
        if history is not None:
            vol_score = self.normalize_score(volatility, history.get('volatility', pd.Series([volatility])))
            mom_score = self.normalize_score(momentum, history.get('momentum', pd.Series([momentum])))
            social_score = self.normalize_score(social_mentions, history.get('social', pd.Series([social_mentions])))
            survey_score = bull_sentiment * 100
            dom_score = self.normalize_score(btc_dominance, history.get('dominance', pd.Series([btc_dominance])))
            search_score = self.normalize_score(google_trends, history.get('search', pd.Series([google_trends])))
        else:
            vol_score = volatility
            mom_score = momentum
            social_score = social_mentions
            survey_score = bull_sentiment * 100
            dom_score = btc_dominance
            search_score = google_trends
        
        # 加权求和
        fgi = (
            self.weights['volatility'] * vol_score +
            self.weights['momentum'] * mom_score +
            self.weights['social'] * social_score +
            self.weights['survey'] * survey_score +
            self.weights['dominance'] * dom_score +
            self.weights['search'] * search_score
        )
        
        # 确定情绪等级
        if fgi < 20:
            level = SentimentLevel.EXTREME_FEAR
        elif fgi < 40:
            level = SentimentLevel.FEAR
        elif fgi < 60:
            level = SentimentLevel.NEUTRAL
        elif fgi < 80:
            level = SentimentLevel.GREED
        else:
            level = SentimentLevel.EXTREME_GREED
        
        return FearGreedIndex(
            timestamp=timestamp,
            value=fgi,
            level=level,
            volatility_score=vol_score,
            momentum_score=mom_score,
            social_score=social_score,
            survey_score=survey_score,
            dominance_score=dom_score,
            search_score=search_score
        )


class RSICalculator:
    """RSI计算器"""
    
    def __init__(self, period: int = 14):
        self.period = period
    
    def calculate(self, prices: pd.Series) -> pd.Series:
        """
        计算RSI
        
        Args:
            prices: 价格序列
        
        Returns:
            RSI序列
        """
        deltas = prices.diff()
        gains = deltas.where(deltas > 0, 0)
        losses = -deltas.where(deltas < 0, 0)
        
        avg_gain = gains.rolling(window=self.period, min_periods=self.period).mean()
        avg_loss = losses.rolling(window=self.period, min_periods=self.period).mean()
        
        rs = avg_gain / avg_loss
        rsi = 100 - (100 / (1 + rs))
        
        return rsi


class CryptoFearGreedStrategy:
    """加密货币恐慌指数交易策略"""
    
    def __init__(
        self,
        buy_threshold: float = 25.0,
        sell_threshold: float = 75.0,
        rsi_buy_threshold: float = 35.0,
        rsi_sell_threshold: float = 65.0,
        zscore_threshold: float = 1.5,
        holding_period: int = 7
    ):
        """
        初始化策略
        
        Args:
            buy_threshold: 买入FGI阈值
            sell_threshold: 卖出FGI阈值
            rsi_buy_threshold: RSI买入阈值
            rsi_sell_threshold: RSI卖出阈值
            zscore_threshold: Z-score阈值
            holding_period: 最大持仓期
        """
        self.buy_threshold = buy_threshold
        self.sell_threshold = sell_threshold
        self.rsi_buy_threshold = rsi_buy_threshold
        self.rsi_sell_threshold = rsi_sell_threshold
        self.zscore_threshold = zscore_threshold
        self.holding_period = holding_period
        
        self.fgi_calculator = CryptoFearGreedIndexCalculator()
        self.rsi_calculator = RSICalculator(period=14)
    
    def calculate_zscore(self, fgi_history: pd.Series, current: float) -> float:
        """计算Z-score偏离"""
        if len(fgi_history) < 30:
            return 0.0
        
        mean = fgi_history.mean()
        std = fgi_history.std()
        
        if std == 0:
            return 0.0
        
        return (current - mean) / std
    
    def detect_divergence(
        self,
        price_history: pd.Series,
        fgi_history: pd.Series
    ) -> Tuple[bool, str]:
        """
        检测价格与FGI背离
        
        Args:
            price_history: 价格历史
            fgi_history: FGI历史
        
        Returns:
            (是否背离, 背离类型)
        """
        if len(price_history) < 20 or len(fgi_history) < 20:
            return False, ""
        
        # 近期最低价
        recent_low_price = price_history.iloc[-20:].min()
        current_price = price_history.iloc[-1]
        price_new_low = current_price <= recent_low_price
        
        # 近期最低FGI
        recent_low_fgi = fgi_history.iloc[-20:].min()
        current_fgi = fgi_history.iloc[-1]
        fgi_new_low = current_fgi <= recent_low_fgi
        
        # 看涨背离：价格新低但FGI未新低
        if price_new_low and not fgi_new_low:
            return True, "bullish_divergence"
        
        # 看跌背离：价格新高但FGI未新高
        recent_high_price = price_history.iloc[-20:].max()
        price_new_high = current_price >= recent_high_price
        
        recent_high_fgi = fgi_history.iloc[-20:].max()
        fgi_new_high = current_fgi >= recent_high_fgi
        
        if price_new_high and not fgi_new_high:
            return True, "bearish_divergence"
        
        return False, ""
    
    def generate_signals(
        self,
        fgi_current: float,
        rsi_current: float,
        zscore_current: float,
        fgi_history: Optional[pd.Series] = None
    ) -> Tuple[str, float]:
        """
        生成交易信号
        
        Args:
            fgi_current: 当前FGI
            rsi_current: 当前RSI
            zscore_current: 当前Z-score
            fgi_history: FGI历史（用于计算偏离）
        
        Returns:
            (信号类型, 置信度)
        """
        # 极度恐惧 + RSI超卖
        if fgi_current < self.buy_threshold and rsi_current < self.rsi_buy_threshold:
            return "BUY", 0.9
        
        # Z-score极端负偏离
        if zscore_current < -self.zscore_threshold:
            return "BUY", 0.8
        
        # 极度贪婪 + RSI超买
        if fgi_current > self.sell_threshold and rsi_current > self.rsi_sell_threshold:
            return "SELL", 0.9
        
        # Z-score极端正偏离
        if zscore_current > self.zscore_threshold:
            return "SELL", 0.8
        
        # 中性区域
        return "HOLD", 0.5
    
    def run_backtest(
        self,
        price_data: pd.DataFrame,
        fgi_data: pd.DataFrame,
        volatility_data: pd.Series,
        social_data: pd.Series
    ) -> pd.DataFrame:
        """
        回测策略
        
        Args:
            price_data: 价格数据
            fgi_data: FGI历史
            volatility_data: 波动率数据
            social_data: 社交媒体数据
        
        Returns:
            回测结果
        """
        results = []
        position = None
        entry_date = None
        entry_price = 0
        
        rsi_series = self.rsi_calculator.calculate(price_data['close'])
        
        for i in range(30, len(price_data)):
            date = price_data['date'].iloc[i]
            close = price_data['close'].iloc[i]
            rsi = rsi_series.iloc[i]
            
            fgi_history = fgi_data['fgi'].iloc[max(0, i-30):i]
            fgi_current = fgi_data['fgi'].iloc[i]
            zscore = self.calculate_zscore(fgi_history, fgi_current)
            
            price_history = price_data['close'].iloc[max(0, i-30):i]
            is_divergence, div_type = self.detect_divergence(price_history, fgi_history)
            
            if position is None:
                signal, confidence = self.generate_signals(
                    fgi_current, rsi, zscore, fgi_history
                )
                
                if signal == "BUY":
                    position = {
                        "type": "LONG",
                        "entry_date": date,
                        "entry_price": close,
                        "entry_fgi": fgi_current,
                        "entry_rsi": rsi
                    }
                    results.append({
                        "date": date,
                        "action": "BUY",
                        "price": close,
                        "fgi": fgi_current,
                        "rsi": rsi,
                        "zscore": zscore,
                        "confidence": confidence,
                        "reason": f"FGI={fgi_current:.1f}, RSI={rsi:.1f}"
                    })
                elif signal == "SELL":
                    position = {
                        "type": "SHORT",
                        "entry_date": date,
                        "entry_price": close,
                        "entry_fgi": fgi_current,
                        "entry_rsi": rsi
                    }
                    results.append({
                        "date": date,
                        "action": "SELL",
                        "price": close,
                        "fgi": fgi_current,
                        "rsi": rsi,
                        "zscore": zscore,
                        "confidence": confidence,
                        "reason": f"FGI={fgi_current:.1f}, RSI={rsi:.1f}"
                    })
            else:
                days_held = (date - entry_date).days
                
                # 计算盈亏
                if position['type'] == "LONG":
                    pnl = (close - entry_price) / entry_price
                else:
                    pnl = (entry_price - close) / entry_price
                
                should_close = False
                close_reason = ""
                
                # 止盈止损
                if pnl > 0.15:
                    should_close = True
                    close_reason = "take_profit"
                elif pnl < -0.05:
                    should_close = True
                    close_reason = "stop_loss"
                elif days_held >= self.holding_period:
                    should_close = True
                    close_reason = "holding_period"
                
                if should_close:
                    results.append({
                        "date": date,
                        "action": f"CLOSE({close_reason})",
                        "price": close,
                        "pnl": pnl,
                        "holding_days": days_held
                    })
                    position = None
                    entry_date = None
        
        return pd.DataFrame(results)


def generate_synthetic_crypto_data(days: int = 252) -> Tuple[pd.DataFrame, pd.DataFrame, pd.Series, pd.Series]:
    """生成合成加密货币数据"""
    np.random.seed(42)
    
    dates = pd.date_range(start='2024-01-01', periods=days, freq='D')
    
    # 价格数据
    btc_price = 40000 * np.exp(np.cumsum(np.random.normal(0.002, 0.04, days)))
    price_df = pd.DataFrame({
        'date': dates,
        'close': btc_price,
        'open': btc_price * (1 + np.random.uniform(-0.01, 0.01, days)),
        'high': btc_price * (1 + np.random.uniform(0, 0.02, days)),
        'low': btc_price * (1 - np.random.uniform(0, 0.02, days)),
        'volume': np.random.randint(10000, 50000, days)
    })
    
    # FGI数据
    fgi_values = np.random.uniform(20, 80, days)
    fgi_df = pd.DataFrame({
        'date': dates,
        'fgi': fgi_values,
        'volatility': np.random.uniform(30, 70, days),
        'momentum': np.random.uniform(30, 70, days),
        'social': np.random.uniform(30, 70, days),
        'survey': np.random.uniform(30, 70, days)
    })
    
    # 波动率数据
    volatility = pd.Series(np.random.uniform(0.02, 0.08, days), index=dates)
    
    # 社交媒体数据
    social = pd.Series(np.random.randint(500, 5000, days), index=dates)
    
    return price_df, fgi_df, volatility, social


if __name__ == "__main__":
    print("=" * 60)
    print("加密货币恐慌指数策略回测")
    print("=" * 60)
    
    # 生成数据
    price_df, fgi_df, volatility, social = generate_synthetic_crypto_data(days=252)
    
    # 创建策略
    strategy = CryptoFearGreedStrategy(
        buy_threshold=25.0,
        sell_threshold=75.0,
        rsi_buy_threshold=35.0,
        rsi_sell_threshold=65.0,
        zscore_threshold=1.5,
        holding_period=7
    )
    
    # 回测
    results = strategy.run_backtest(price_df, fgi_df, volatility, social)
    
    if len(results) > 0:
        print("\n【回测结果】")
        print(results)
        
        close_results = results[results['action'].str.startswith('CLOSE')]
        if len(close_results) > 0:
            print(f"\n平均持仓天数: {close_results['holding_days'].mean():.1f}")
            print(f"平均收益率: {close_results['pnl'].mean()*100:.2f}%")
            print(f"胜率: {(close_results['pnl'] > 0).mean()*100:.1f}%")
    else:
        print("\n未产生交易信号")
